### ISSUE TITLE
Fail to change Service offering from local <> shared storage

### ISSUE THREAD

**Issue by GabrielBrascher** [apache/cloudstack#4545]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
it's there for a long time (+9 years)
~~~

##### CONFIGURATION

Cluster with local and shared storage pools

##### SUMMARY

Migrated a volume from local to shared storage pool; however, it fails to scale the offering from local to shared.

Error: "Unable to upgrade virtual machine VM [User|i-2-345-VM], cannot switch between local storage and shared storage service offerings.  Current offering useLocalStorage=true, target offering useLocalStorage=false"

Caused due to code at: https://github.com/apache/cloudstack/blob/58a0a7b1a3408240757279a95543ce6304cd2a62/engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java#L3647

~~~
        // Check that the service offering being upgraded to has the same storage pool preference as the VM's current service
        // offering
        if (currentServiceOffering.isUseLocalStorage() != newServiceOffering.isUseLocalStorage()) {
            throw new InvalidParameterValueException("Unable to upgrade virtual machine " + vmInstance.toString() +
                    ", cannot switch between local storage and shared storage service offerings.  Current offering " + "useLocalStorage=" +
                    currentServiceOffering.isUseLocalStorage() + ", target offering useLocalStorage=" + newServiceOffering.isUseLocalStorage());
        }
~~~

##### STEPS TO REPRODUCE

~~~
1. Stop VM
2. Migrate VM's volume from local to shared primary storage
3. Update service offering
4. Start VM
~~~

##### EXPECTED RESULTS

~~~
VM gets migrated from local to shared storage
~~~

##### ACTUAL RESULTS

~~~
Fail to update service offering from local to shared storage; therefore, it fails to start the VM with the migrated volume
~~~

**Comment by aleskxyz** [comment]
@GabrielBrascher 
Is it possible to extend this issue to changing disk offering for different tags?
For example I want to migrate a disk from HDD to SSD.

**Comment by GabrielBrascher** [comment]
@aleskxyz I think that PR https://github.com/apache/cloudstack/pull/4283 already does what you are looking for.

Initially, PR https://github.com/apache/cloudstack/pull/2636 (Merged) had it implemented but later this was changed; PR #4283  (Open) proposes a code to get that functionality back.

**Comment by DaanHoogland** [comment]
@GabrielBrascher by the looks of the line in your description this is very intentional. Do you plane to propose to change it?
And do you want to implement the possibility two way?

**Comment by GabrielBrascher** [comment]
@DaanHoogland I think that the validation was intentional indeed, but it does not consider that it is possible to migrate Volumes between local <> shared storage pools.

The conditional checks the previous offering, which works really well if the volume has not been migrated; however, if the volume has been moved to a storage pool with different scope (e.g. HOST or Cluster) than the offering (e.g. local or shared) then this can be a problem.

To bring some context, here is an example.
1. for now, it is possible to migrate a VM's  Root disk from a shared storage pool to local storage;
2. however, it is not possible to upgrade its offering to an offering that matches the new storage pool that fits the new pool because the validation offering upgrade command does not allow to do it so.

In my opinion, there should not exist such a limitation as the migration is perfectly doable; and also very necessary in some cases while managing Volumes across storage pools for maintenance or other purposes.

For now, the only way of doing it without a fix is by changing the Volume and VM's offerings on DB.

**Comment by GabrielBrascher** [comment]
Update: working to fix this issue via PR #4915

**Comment by rhtyd** [comment]
Fixed in https://github.com/apache/cloudstack/pull/4915
